File: sdks/go/middleware/conneg.go
// Package middleware provides HTTP serving for `.cv` resources with content
// negotiation that mirrors @cvfile/server (Node) and cvfile.server (Python)
// exactly. The same Accept / ?format= request gets the same representation
// regardless of the implementing language.
package middleware

import (
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// ServeFormat is the chosen representation: pdf, markdown, or html.
type ServeFormat string

const (
	FormatPDF      ServeFormat = "pdf"
	FormatMarkdown ServeFormat = "markdown"
	FormatHTML     ServeFormat = "html"

	// PDFPrimaryMIME is the cv-strict media type advertised in Link headers.
	PDFPrimaryMIME = "application/vnd.cv+pdf"
	// PDFFallbackMIME is the bare PDF type we accept on inbound Accept.
	PDFFallbackMIME = "application/pdf"
)

// NegotiationResult is the chosen format plus the language preference (if any).
type NegotiationResult struct {
	Format   ServeFormat
	Language string
}

// NegotiateInput is the raw HTTP-derived inputs.
type NegotiateInput struct {
	Accept         string
	AcceptLanguage string
	FormatQuery    string
}

var formatByMIME = map[string]ServeFormat{
	"text/markdown":          FormatMarkdown,
	"text/x-markdown":        FormatMarkdown,
	"text/html":              FormatHTML,
	"application/xhtml+xml":  FormatHTML,
	"application/pdf":        FormatPDF,
	"application/vnd.cv+pdf": FormatPDF,
}

var formatByQuery = map[string]ServeFormat{
	"md":       FormatMarkdown,
	"markdown": FormatMarkdown,
	"html":     FormatHTML,
	"pdf":      FormatPDF,
	"cv":       FormatPDF,
}

var qRE = regexp.MustCompile(`(?i)^q\s*=\s*(\d*\.?\d+)`)

type acceptEntry struct {
	Type string
	Q    float64
}

// ParseAccept returns Accept entries sorted by descending q-value.
func ParseAccept(header string) []acceptEntry {
	if header == "" {
		return nil
	}
	var out []acceptEntry
	for _, raw := range strings.Split(header, ",") {
		bits := strings.Split(strings.TrimSpace(raw), ";")
		for i := range bits {
			bits[i] = strings.TrimSpace(bits[i])
		}
		if len(bits) == 0 || bits[0] == "" {
			continue
		}
		q := 1.0
		for _, p := range bits[1:] {
			if m := qRE.FindStringSubmatch(p); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					q = v
				}
			}
		}
		out = append(out, acceptEntry{Type: strings.ToLower(bits[0]), Q: q})
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].Q > out[j].Q })
	return out
}

// ParseAcceptLanguage returns language tags sorted by descending q-value,
// stripping any wildcard "*" entries.
func ParseAcceptLanguage(header string) []string {
	if header == "" {
		return nil
	}
	type pair struct {
		tag string
		q   float64
	}
	var pairs []pair
	for _, raw := range strings.Split(header, ",") {
		bits := strings.Split(strings.TrimSpace(raw), ";")
		for i := range bits {
			bits[i] = strings.TrimSpace(bits[i])
		}
		if len(bits) == 0 || bits[0] == "" {
			continue
		}
		tag := strings.ToLower(bits[0])
		if tag == "*" {
			continue
		}
		q := 1.0
		for _, p := range bits[1:] {
			if m := qRE.FindStringSubmatch(p); len(m) > 1 {
				if v, err := strconv.ParseFloat(m[1], 64); err == nil {
					q = v
				}
			}
		}
		pairs = append(pairs, pair{tag: tag, q: q})
	}
	sort.SliceStable(pairs, func(i, j int) bool { return pairs[i].q > pairs[j].q })
	tags := make([]string, len(pairs))
	for i, p := range pairs {
		tags[i] = p.tag
	}
	return tags
}

// Negotiate picks a format based on ?format=, then Accept, defaulting to PDF.
func Negotiate(in NegotiateInput) NegotiationResult {
	languages := ParseAcceptLanguage(in.AcceptLanguage)
	var lang string
	if len(languages) > 0 {
		lang = languages[0]
	}

	if in.FormatQuery != "" {
		if f, ok := formatByQuery[strings.ToLower(in.FormatQuery)]; ok {
			return NegotiationResult{Format: f, Language: lang}
		}
	}

	for _, a := range ParseAccept(in.Accept) {
		if direct, ok := formatByMIME[a.Type]; ok {
			return NegotiationResult{Format: direct, Language: lang}
		}
		if a.Type == "*/*" || a.Type == "application/*" {
			return NegotiationResult{Format: FormatPDF, Language: lang}
		}
		if a.Type == "text/*" {
			return NegotiationResult{Format: FormatHTML, Language: lang}
		}
	}

	return NegotiationResult{Format: FormatPDF, Language: lang}
}

// BuildLinkHeader composes the alternates Link header advertised on every
// `.cv` response so well-behaved clients can discover the markdown/html paths.
func BuildLinkHeader(selfURL string) string {
	sep := "?"
	if strings.Contains(selfURL, "?") {
		sep = "&"
	}
	return strings.Join([]string{
		`<` + selfURL + `>; rel="alternate"; type="` + PDFPrimaryMIME + `"`,
		`<` + selfURL + sep + `format=md>; rel="alternate"; type="text/markdown"`,
		`<` + selfURL + sep + `format=html>; rel="alternate"; type="text/html"`,
	}, ", ")
}

File: sdks/python/src/cvfile/server/_conneg.py
"""Content negotiation for `.cv` resources.

Mirrors @cvfile/server's conneg algorithm exactly. ?format=md|html|pdf wins
over Accept; otherwise q-sorted Accept; defaults to PDF.
"""

from __future__ import annotations

import contextlib
import re
from dataclasses import dataclass
from typing import Literal

ServeFormat = Literal["pdf", "markdown", "html"]

PDF_PRIMARY_MIME = "application/vnd.cv+pdf"
PDF_FALLBACK_MIME = "application/pdf"

_FORMAT_BY_MIME: dict[str, ServeFormat] = {
    "text/markdown": "markdown",
    "text/x-markdown": "markdown",
    "text/html": "html",
    "application/xhtml+xml": "html",
    "application/pdf": "pdf",
    "application/vnd.cv+pdf": "pdf",
}

_FORMAT_BY_QUERY: dict[str, ServeFormat] = {
    "md": "markdown",
    "markdown": "markdown",
    "html": "html",
    "pdf": "pdf",
    "cv": "pdf",
}

_Q_RE = re.compile(r"^q\s*=\s*(\d*\.?\d+)", re.IGNORECASE)


@dataclass(frozen=True, slots=True)
class NegotiationResult:
    format: ServeFormat
    language: str | None


@dataclass(frozen=True, slots=True)
class _ParsedAccept:
    type: str
    q: float


def parse_accept(header: str | None) -> list[_ParsedAccept]:
    if not header:
        return []
    parts: list[_ParsedAccept] = []
    for raw in header.split(","):
        bits = [b.strip() for b in raw.strip().split(";")]
        if not bits or not bits[0]:
            continue
        q = 1.0
        for p in bits[1:]:
            m = _Q_RE.match(p)
            if m:
                with contextlib.suppress(ValueError):
                    q = float(m.group(1))
        parts.append(_ParsedAccept(type=bits[0].lower(), q=q))
    parts.sort(key=lambda p: p.q, reverse=True)
    return parts


def parse_accept_language(header: str | None) -> list[str]:
    if not header:
        return []
    pairs: list[tuple[str, float]] = []
    for raw in header.split(","):
        bits = [b.strip() for b in raw.strip().split(";")]
        if not bits or not bits[0]:
            continue
        tag = bits[0].lower()
        if tag == "*":
            continue
        q = 1.0
        for p in bits[1:]:
            m = _Q_RE.match(p)
            if m:
                with contextlib.suppress(ValueError):
                    q = float(m.group(1))
        pairs.append((tag, q))
    pairs.sort(key=lambda p: p[1], reverse=True)
    return [tag for tag, _ in pairs]


def negotiate(
    *,
    accept: str | None = None,
    accept_language: str | None = None,
    format_query: str | None = None,
) -> NegotiationResult:
    languages = parse_accept_language(accept_language)
    language = languages[0] if languages else None

    if format_query:
        from_query = _FORMAT_BY_QUERY.get(format_query.lower())
        if from_query:
            return NegotiationResult(format=from_query, language=language)

    for entry in parse_accept(accept):
        direct = _FORMAT_BY_MIME.get(entry.type)
        if direct:
            return NegotiationResult(format=direct, language=language)
        if entry.type in ("*/*", "application/*"):
            return NegotiationResult(format="pdf", language=language)
        if entry.type == "text/*":
            return NegotiationResult(format="html", language=language)

    return NegotiationResult(format="pdf", language=language)


def build_link_header(self_url: str, *, cv_mime: str = PDF_PRIMARY_MIME) -> str:
    sep = "&" if "?" in self_url else "?"
    return ", ".join(
        [
            f'<{self_url}>; rel="alternate"; type="{cv_mime}"',
            f'<{self_url}{sep}format=md>; rel="alternate"; type="text/markdown"',
            f'<{self_url}{sep}format=html>; rel="alternate"; type="text/html"',
        ]
    )
